// redux-first-routing/lib/actions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.locationChange = exports.goForward = exports.goBack = exports.go = exports.replace = exports.push = undefined;

var _queryString = require('query-string');

var _constants = require('./constants');

var push = exports.push = function push(href) {
  return {
    type: _constants.PUSH,
    payload: href
  };
};

var replace = exports.replace = function replace(href) {
  return {
    type: _constants.REPLACE,
    payload: href
  };
};

var go = exports.go = function go(index) {
  return {
    type: _constants.GO,
    payload: index
  };
};

var goBack = exports.goBack = function goBack() {
  return {
    type: _constants.GO_BACK
  };
};

var goForward = exports.goForward = function goForward() {
  return {
    type: _constants.GO_FORWARD
  };
};

var locationChange = exports.locationChange = function locationChange(_ref) {
  var pathname = _ref.pathname,
      search = _ref.search,
      hash = _ref.hash;
  return {
    type: _constants.LOCATION_CHANGE,
    payload: {
      pathname: pathname,
      search: search,
      queries: (0, _queryString.parse)(search),
      hash: hash
    }
  };
};
